File: backend/internal/service/update.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"

	"golang.org/x/mod/semver"
)

const (
	defaultReleaseURL = "https://api.github.com/repos/nasnet-community/nasnet-monitor/releases/latest"
	updateCacheTTL    = 6 * time.Hour
	updateFetchLimit  = 10 * time.Second
)

type UpdateCheck struct {
	CurrentVersion  string `json:"currentVersion"`
	LatestVersion   string `json:"latestVersion,omitempty"`
	UpdateAvailable bool   `json:"updateAvailable"`
	ReleaseURL      string `json:"releaseUrl,omitempty"`
}

type UpdateService struct {
	version    string
	releaseURL string
	client     *http.Client

	mu        sync.Mutex
	cached    *UpdateCheck
	fetchedAt time.Time
}

func NewUpdateService(version string) *UpdateService {
	return &UpdateService{
		version:    version,
		releaseURL: defaultReleaseURL,
		client:     &http.Client{Timeout: updateFetchLimit},
	}
}

func (s *UpdateService) Check(ctx context.Context) (*UpdateCheck, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.cached != nil && time.Since(s.fetchedAt) < updateCacheTTL {
		return s.cached, nil
	}

	result, err := s.fetch(ctx)
	if err != nil {
		return nil, err
	}
	s.cached = result
	s.fetchedAt = time.Now()
	return result, nil
}

type githubRelease struct {
	TagName string `json:"tag_name"`
	HTMLURL string `json:"html_url"`
}

func (s *UpdateService) fetch(ctx context.Context) (*UpdateCheck, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.releaseURL, http.NoBody)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/vnd.github+json")

	res, err := s.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("fetching latest release: %w", err)
	}
	defer func() { _ = res.Body.Close() }()

	// 404 means the repo has no (non-prerelease) releases yet.
	if res.StatusCode == http.StatusNotFound {
		return &UpdateCheck{CurrentVersion: s.version}, nil
	}
	if res.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("github returned %d", res.StatusCode)
	}

	var release githubRelease
	if err := json.NewDecoder(res.Body).Decode(&release); err != nil {
		return nil, fmt.Errorf("decoding release: %w", err)
	}

	latest := strings.TrimPrefix(release.TagName, "v")
	return &UpdateCheck{
		CurrentVersion:  s.version,
		LatestVersion:   latest,
		UpdateAvailable: newerVersion(s.version, latest),
		ReleaseURL:      release.HTMLURL,
	}, nil
}

func newerVersion(current, latest string) bool {
	c, l := "v"+current, "v"+latest
	if !semver.IsValid(c) || !semver.IsValid(l) {
		return false
	}
	return semver.Compare(l, c) > 0
}

File: backend/src/update.rs
use std::sync::Mutex;
use std::time::{Duration, Instant};

use serde::{Deserialize, Serialize};

const DEFAULT_RELEASE_URL: &str =
    "https://api.github.com/repos/nasnet-community/nasnet-monitor/releases/latest";
const UPDATE_CACHE_TTL: Duration = Duration::from_secs(6 * 60 * 60);
const UPDATE_FETCH_LIMIT: Duration = Duration::from_secs(10);

#[derive(Debug, thiserror::Error)]
pub enum UpdateError {
    #[error("fetching latest release: {0}")]
    Fetch(String),
    #[error("github returned {0}")]
    UnexpectedStatus(u16),
    #[error("decoding release: {0}")]
    Decode(String),
    #[error("update check task: {0}")]
    Task(#[from] tokio::task::JoinError),
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateCheck {
    pub current_version: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub latest_version: String,
    pub update_available: bool,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub release_url: String,
}

impl UpdateCheck {
    fn none(version: &str) -> Self {
        Self {
            current_version: version.to_string(),
            latest_version: String::new(),
            update_available: false,
            release_url: String::new(),
        }
    }
}

/// Checks GitHub for a newer release, caching the result for six hours.
pub struct UpdateService {
    version: &'static str,
    release_url: String,
    cache: Mutex<Option<(Instant, UpdateCheck)>>,
}

#[derive(Deserialize)]
struct GithubRelease {
    #[serde(default)]
    tag_name: String,
    #[serde(default)]
    html_url: String,
}

impl UpdateService {
    pub fn new(version: &'static str) -> Self {
        Self {
            version,
            // Overridable for testing against a local mock.
            release_url: std::env::var("RELEASE_URL")
                .unwrap_or_else(|_| DEFAULT_RELEASE_URL.to_string()),
            cache: Mutex::new(None),
        }
    }

    pub async fn check(&self) -> Result<UpdateCheck, UpdateError> {
        if let Some((fetched_at, cached)) = self.cache.lock().unwrap().as_ref() {
            if fetched_at.elapsed() < UPDATE_CACHE_TTL {
                return Ok(cached.clone());
            }
        }

        let version = self.version;
        let url = self.release_url.clone();
        // ureq is a blocking client; run it off the async runtime.
        let result = tokio::task::spawn_blocking(move || fetch(&url, version)).await??;

        *self.cache.lock().unwrap() = Some((Instant::now(), result.clone()));
        Ok(result)
    }
}

fn fetch(url: &str, version: &str) -> Result<UpdateCheck, UpdateError> {
    let agent = ureq::AgentBuilder::new()
        .timeout(UPDATE_FETCH_LIMIT)
        .build();
    let response = agent
        .get(url)
        .set("Accept", "application/vnd.github+json")
        .call();

    let release: GithubRelease = match response {
        Ok(res) => res
            .into_json()
            .map_err(|e| UpdateError::Decode(e.to_string()))?,
        // 404 means the repo has no (non-prerelease) releases yet.
        Err(ureq::Error::Status(404, _)) => return Ok(UpdateCheck::none(version)),
        Err(ureq::Error::Status(code, _)) => return Err(UpdateError::UnexpectedStatus(code)),
        Err(e) => return Err(UpdateError::Fetch(e.to_string())),
    };

    let latest = release.tag_name.trim_start_matches('v').to_string();
    Ok(UpdateCheck {
        current_version: version.to_string(),
        update_available: newer_version(version, &latest),
        latest_version: latest,
        release_url: release.html_url,
    })
}

/// Whether `latest` is a strictly newer semver than `current`. Unparseable
/// versions never report an update, matching the Go backend.
fn newer_version(current: &str, latest: &str) -> bool {
    match (
        semver::Version::parse(current),
        semver::Version::parse(latest),
    ) {
        (Ok(current), Ok(latest)) => latest > current,
        _ => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn newer_version_compares_semver() {
        assert!(newer_version("0.1.0", "0.2.0"));
        assert!(newer_version("0.1.0", "1.0.0"));
        assert!(!newer_version("0.2.0", "0.2.0"));
        assert!(!newer_version("0.3.0", "0.2.9"));
    }

    #[test]
    fn newer_version_rejects_invalid_input() {
        assert!(!newer_version("not-a-version", "0.2.0"));
        assert!(!newer_version("0.1.0", "latest"));
    }

    #[test]
    fn update_check_omits_empty_fields() {
        let body = serde_json::to_string(&UpdateCheck::none("0.1.0")).unwrap();
        assert_eq!(
            body,
            "{\"currentVersion\":\"0.1.0\",\"updateAvailable\":false}"
        );
    }
}
